tools/backup_analysis.py

```diff
@@ -0,0 +1,175 @@
+
+def usage_exit():
+    print('Usage: backup_analysis.py dir1 dir2 [granularity]')
+    print('  dir1 and dir2 are POSIX pathnames to WiredTiger backup directories,')
+    print('Options:')
+    print('  granularity - an (optional) positive integer indicating the granularity')
+    print('  of the incremental backup blocks. The default is 16MB')
+    sys.exit(1)
+
+def die(reason):
+    print('backup_analysis.py: error: ' + reason, file=sys.stderr)
+    sys.exit(1)
+
+# Check that the directory given is a backup. Look for the WiredTiger.backup file.
+def check_backup(mydir):
+    if not os.path.isdir(mydir):
+        return False
+    backup_file = mydir + "/WiredTiger.backup"
+    if not os.path.exists(backup_file):
+        return False
+    return True
+
+def compare_file(dir1, dir2, filename, file_size, cmp_size, granularity):
+    # Initialize all of our counters per file.
+    bytes_gran = 0
+    gran_blocks = 0
+    num_blocks = file_size // cmp_size
+    partial = file_size % cmp_size
+    pct20 = granularity // 5
+    pct80 = pct20 * 4
+    pct20_count = 0
+    pct80_count = 0
+    start_off = 0
+    total_bytes_diff = 0
+    fp1 = open(os.path.join(dir1, filename), "rb")
+    fp2 = open(os.path.join(dir2, filename), "rb")
+    # Time how long it takes to compare each file.
+    start = time.asctime()
+    # Compare the bytes in cmp_size blocks between both files.
+    print("")
+    for b in range(0, num_blocks + 1):
+        offset = cmp_size * b
+        # Gather and report information when we cross a granularity boundary.
+        if offset % granularity == 0:
+            if offset != 0 and bytes_gran != 0:
+                print(f'{filename}: offset {start_off}: {bytes_gran} bytes differ in {granularity} bytes')
+            # Account for small or large block changes.
+            if bytes_gran != 0:
+                if bytes_gran <= pct20:
+                    pct20_count += 1
+                elif bytes_gran >= pct80:
+                    pct80_count += 1
+            # Reset for the next granularity block.
+            start_off = offset
+            bytes_gran = 0
+
+        # Compare the two blocks. We know both files are at least file_size so all reads should work.
+        buf1 = fp1.read(cmp_size)
+        buf2 = fp2.read(cmp_size)
+        # If they're different, gather information.
+        if buf1 != buf2:
+            total_bytes_diff += cmp_size
+            # Count how many granularity level blocks changed.
+            if bytes_gran == 0:
+                gran_blocks += 1
+            bytes_gran += cmp_size
+    # Account for anything from the final iteration of the loop and partial blocks.
+    if partial != 0:
+        buf1 = fp1.read(partial)
+        buf2 = fp2.read(partial)
+        # If they're different, gather information.
+        if buf1 != buf2:
+            total_bytes_diff += partial
+            bytes_gran += partial
+    fp1.close()
+    fp2.close()
+    end = time.asctime()
+    if bytes_gran != 0:
```

Comment:
It seems like this block would more logically belong to the `if partial != 0:` block just before you close the file?

Reply:
I have rearranged the code to only have one conditional about the per-granularity-block output.